common/ast/src/parser/query.rs

```diff
@@ -311,3 +284,190 @@ pub fn order_by_expr(i: Input) -> IResult<OrderByExpr> {
+        group_by: Vec<Expr<'a>>,
+        having: Option<Expr<'a>>,
+    },
+    SetOperation {
+        span: &'a [Token<'a>],
+        op: SetOperator,
+        all: bool,
+    },
+    Group(SetExpr<'a>),
+}
+
+pub fn set_operation_element(i: Input) -> IResult<SetOperationElement> {
+    let set_operator = map(
+        consumed(rule! {
+            ( UNION | EXCEPT | INTERSECT ) ~ ALL?
+        }),
+        |(span, (op, all))| {
+            let op = match op.kind {
+                UNION => SetOperator::Union,
+                INTERSECT => SetOperator::Intersect,
+                EXCEPT => SetOperator::Except,
+                _ => unreachable!(),
+            };
+            SetOperationElement::SetOperation {
+                span: span.0,
+                op,
+                all: all.is_some(),
+            }
+        },
+    );
+
+    let select_stmt = map(
+        consumed(rule! {
+             SELECT ~ DISTINCT? ~ ^#comma_separated_list1(select_target)
+                ~ ( FROM ~ ^#comma_separated_list1(table_reference) )?
+                ~ ( WHERE ~ ^#expr )?
+                ~ ( GROUP ~ ^BY ~ ^#comma_separated_list1(expr) )?
+                ~ ( HAVING ~ ^#expr )?
+        }),
+        |(
+            span,
+            (
+                _select,
+                opt_distinct,
+                select_list,
+                opt_from_block,
+                opt_where_block,
+                opt_group_by_block,
+                opt_having_block,
+            ),
+        )| {
+            SetOperationElement::SelectStmt {
+                span: span.0,
+                distinct: opt_distinct.is_some(),
+                select_list,
+                from: opt_from_block
+                    .map(|(_, table_refs)| table_refs)
+                    .unwrap_or_default(),
+                selection: opt_where_block.map(|(_, selection)| selection),
+                group_by: opt_group_by_block
+                    .map(|(_, _, group_by)| group_by)
+                    .unwrap_or_default(),
+                having: opt_having_block.map(|(_, having)| having),
+            }
+        },
+    );
+
+    let group = map(
+        rule! {
+           "("
+           ~ ^#set_operation
+           ~ ^")"
+        },
+        |(_, set_expr, _)| SetOperationElement::Group(set_expr),
+    );
+
+    rule!( #group | #set_operator | #select_stmt)(i)
+}
+
+pub fn set_operation(i: Input) -> IResult<SetExpr> {
+    let (rest, set_operation_elements) = rule!(#set_operation_element+)(i)?;
+    let iter = &mut set_operation_elements.into_iter();
+    let mut iter = iter.peekable();
+    let set_expr = SetOperationParser
+        .parse_input(&mut iter, Precedence(0))
+        .map_err(|err| Error::from_error_kind(rest.slice(..1), ErrorKind::from(err)))
+        .map_err(nom::Err::Error)?;
+    if let Some(_) = iter.peek() {
+        // Rollback parsing footprint on unused expr elements.
+        i.1.clear();
+        Ok((i.slice(1..), set_expr))
```

Comment:
The slice index should be starting from the span of the first unused element. Seems that `WithSpan` is still necessary.

Reply:
Yes, fixed